presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -256,37 +282,32 @@ private void processNewSources()
 
     public ListenableFuture<?> process()
     {
+        checkState(initialized.get(), "Driver is not initialized");
         checkLockNotHeld("Can not process while holding the driver lock");
 
+        // if the driver is blocked we don't need to continue
+        SettableFuture<?> future = driverBlockedFuture.get();
+        if (!future.isDone()) {
+            return future;
+        }
+
         Optional<ListenableFuture<?>> result = tryWithLock(100, TimeUnit.MILLISECONDS, () -> {
-            ListenableFuture<?> driverBlockedFuture = processInternal();
-            return wakeUpOnRevokeRequest(driverBlockedFuture);
+            ListenableFuture<?> sourceBlockedFuture = processInternal();
+            updateDriverBlockedFuture(sourceBlockedFuture);
+            return driverBlockedFuture.get();
```

Comment:
If `updateDriverBlockedFuture()` returned the newly created future, you wouldn't need to re-read it here.